test/gemm-microkernel-tester.cc

```diff
@@ -1230,6 +1242,171 @@ void GemmMicrokernelTester::Test(
+      xnn_init_qs8_qc8w_scale_fp32_params(
+        n(), nr(), nr(),
+        stride,
+        stride,
+        0,
+        bias.data(),
+        (void*) start);
+    }
+
+    // Compute 32-bit results and output quantization arguments.
+    std::fill(c_ref.begin(), c_ref.end(), 0);
+    for (size_t m_index = 0; m_index < m(); m_index++) {
+      for (size_t n_index = 0; n_index < n(); n_index++) {
+        float kfsum = 0.0;
+        for (size_t bl_index=0; bl_index < num_blocks; ++bl_index) {
+          int32_t ksum = 0;
+          int32_t c_ref_acc = 0;
+          for (size_t kr_index = 0; kr_index < bl(); kr_index++) {
+            const size_t k_index =  bl_index * bl() + kr_index;
+            const size_t nb_index = (n_index * k2 + k_index) / 2;
+            const int32_t bv = int32_t((k_index % 2 == 0) ? (b[nb_index] & UINT8_C(0xF)) : (b[nb_index] >> 4)) - b_zero_point();
+            ksum += bv;
+            c_ref_acc += int32_t(a[m_index * a_stride() + k_index]) * int32_t(bv);
+          }
+          size_t scale_index = n_index * num_blocks + bl_index;
+          float scale = kernel_scale2d[scale_index];
+          c_ref[m_index * n() + n_index] += c_ref_acc * scale;
+          kfsum += scale * ksum;
+        }
+        c_ref[m_index * n() + n_index] -= (quantization_params[m_index].zero_point * kfsum);
+        c_ref[m_index * n() + n_index] *= quantization_params[m_index].inv_scale;
+        c_ref[m_index * n() + n_index] += bias[n_index];
+      }
+    }
+
+    const float accumulated_min = *std::min_element(c_ref.cbegin(), c_ref.cend());
+    const float accumulated_max = *std::max_element(c_ref.cbegin(), c_ref.cend());
+    const float c_min =
+        qmin() == std::numeric_limits<uint8_t>::min() ? -std::numeric_limits<float>::infinity()
+        : accumulated_min + (accumulated_max - accumulated_min) / 255.0f * float(qmin());
+    const float c_max =
+        qmax() == std::numeric_limits<uint8_t>::max() ? +std::numeric_limits<float>::infinity()
+        : accumulated_max - (accumulated_max - accumulated_min) / 255.0f * float(255 - qmax());
+
+    // Prepare parameters.
+    xnn_f16_qc4w_minmax_params params;
+    init_params(&params,
+      fp16_ieee_from_fp32_value(c_min),
+      fp16_ieee_from_fp32_value(c_max), 8);
+
+    for (size_t m_index = 0; m_index < m(); m_index++) {
+      for (size_t n_index = 0; n_index < n(); n_index++) {
+        c_ref[m_index * n() + n_index] = std::max(std::min(c_ref[m_index * n() + n_index], c_max), c_min);
+      }
+    }
+
+    gemm(m(), n(), k2, bl(),
+        a.data(), a_stride() * sizeof(int8_t),
+        static_cast<const void*>(packed_w.data()),
+        c.data(), cm_stride() * sizeof(uint16_t), cn_stride() * sizeof(uint16_t), &params, quantization_params.data());
+
+    for (size_t i = 0; i < m(); i++) {
+      for (size_t j = 0; j < n(); j++) {
+        // Extract tolerance into variable to workaround test failures on Linux AArch64.
```

Comment:
Can you dig into what is the minimum tolerance we can get away with for fp16 on Armv8? and on x86? This is something I added pseudo-randomly while doing the first development.

Reply:
For scalar I do expect a tighter tolerance